feat(devtools): create profiler for injector events in dev mode

AleksanderBodurri opened the pull request:
See individual commits

AleksanderBodurri committed:
feat(devtools): create profiler for DI and injector events

Currently, understanding dependency injection in Angular requires a lot of context and has been sited in our surveys as one of the largest points of confusion that our users have with the framework. This commit is the beginning of our approach to make debugging dependency injection in Angular easier.

This commit introduces injector profiler callbacks in parts of the framework to emit injector events. It also introduces a default handler for these events. This default handler parses the stream of events to construct some data structures that will support new injector debug APIs.

We have implemented a similar pattern in the past to minimize overhead. There is also the possiblity of making the internal `setInjectorProfiler` function a public debug API in the future, so that users can implement their own handlers to debug DI events. Lastly DI in Angular maps nicely to a stream of events.

For production applications there is no runtime overhead. For applications in dev mode there is some additional overhead from the default profiler handling injector events and holding debug data in memory.

For production applications, dead code elimination should strip all of the code used by this PR.

AleksanderBodurri committed:
refactor(core): convert walkProviderTree and `processInjectorTypesWithProviders` to callback style interfaces

walkProviderTree and processInjectorTypesWithProviders both perform some generic traversal logic of the import graph of an input NgModule or Standalone component. Currently, these functions pass around a `providersOut` array that is used to collect providers at each step of the traversal.

This PR converts those functions to accept visitor callbacks instead of the `providersOut` array. This is done to make the traversal logic of these functions reusable, while leaving it up to the visitor to determine the logic that fires for each visited node.

This refactor would allow us to reuse `walkProviderTree` for injector debugging APIs that could support some cool features in DevTools, like tracing the injector resolution path of an injected property on a component instance all the way up to the specific imported module/standalone component.

AleksanderBodurri committed:
feat(core): create injector debugging APIs

This commit introduces 3 new APIs.

getDependenciesFromInstantiation:
- Given an injector and a token that was instantiated by that injector, discover all of the things were injected in the instance of that token
- This API is meant to enable recursive inspection of dependencies. Dependencies returned by this API include which injector they were providedIn, which enables the continous use of getDependenciesFromInstantiation to determine the dependencies of dependencies

getInjectorProviders:
- Given an injector, discover all of the providers that were configured in that injector.
- This API returns information on the configured providers of an injector, including the import path that leads to the container that the provider originated from (NgModule or standalone directive). This enables fine grained inspection to determine where a specific provider comes from.

getInjectorParent:
- Given an injector, discover the parent of that injector.
- This function is meant to be used recursively to discover the entire resolution path from a starting injector to the NullInjector.

These APIs were designed to be used together. For example, getInjectorParent can be used to discover the structure of an injector hierarchy. Once that's done, getInjectorProviders can be used to determine the providers of each injector in that hierarchy.

Another example: getDependenciesFromInstantiation can be used to discover the dependencies of a specific injector constructed instance. From there, we can use getInjectorParent to discover the injector resolution path and map each dependency to a path from the starting injector to the injector that it was provided in.

AndrewKushnir commented on packages/core/src/render3/di_setup.ts:
```diff
@@ -74,10 +76,22 @@ function resolveProvider(
 
-    const tNode = getCurrentTNode()!;
+    let providerFactory: () => any;
+    if (typeof ngDevMode === 'undefined' || ngDevMode) {
+      const nodeInjector =
+          new NodeInjector(tNode as TElementNode | TContainerNode | TElementContainerNode, lView);
+      const prevInjectContext = setDebugInjectContext(
+          {injector: new DebugNodeInjector(nodeInjector), token: token as Type<unknown>});
+      providerFactory = providerToFactory(provider);
+      emitProviderConfiguredEvent(provider);
+      setDebugInjectContext(prevInjectContext);
+    } else {
+      providerFactory = providerToFactory(provider);
+    }
```
In cases like this (where the call to instantiate a provider is close to the "wrapping" code to establish debug DI context), it'd be great to extract out all the logic into a separate fn (within `injector_profiler.ts`), so that it's easier to follow and easier to separate what is dev mode and what is not. Here is a quick example (name of the function and arguments - TBD, this is just to illustrate the idea):
```suggestion
    if (typeof ngDevMode === 'undefined' || ngDevMode) {
      // Create a factory in a debug context in dev mode to provide
      // more information to the injector profiler (which works in dev mode only).
      runInDebugInjectContext(tNode, lView, token, provider, () => {
        providerFactory = providerToFactory(provider);
      });
    } else {
      providerFactory = providerToFactory(provider);
    }
```

AndrewKushnir replied:
I'd really love to see this pattern in all other cases as well (and there might be separate fns for NodeInjector and Module/EnvironmentInjector), but we can do it as a followup as well.

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,121 @@
+  standaloneInjectorToComponent: new WeakMap<Injector, Type<unknown>>()
+};
+
+export function setupFrameworkInjectorProfiler(): void {
```
This function would benefit from JSDoc that would contain information about the architecture (some info from the design doc).

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,121 @@
+import {LView} from './interfaces/view';
+
+export const frameworkDIDebugData = {
+  injectorToInstantiatedTokenToDependencies:
+      new WeakMap<Injector|LView, WeakMap<Type<unknown>, InjectedService[]>>(),
+  instanceToInjector: new WeakMap<object, Injector>(),
+  injectorToProviders: new WeakMap<Injector|LView, ProviderRecord[]>(),
+  standaloneInjectorToComponent: new WeakMap<Injector, Type<unknown>>()
```
This object would require JSDoc as well (for the object itself as well as for each field). We'd also need to mention why do we use this global storage (why it's ok in case of multiple components on a page, when do we reset/cleanup, why it can't be a source of memory leaks, etc).

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,121 @@
+    if (type === InjectorProfilerEventType.Inject) {
+      const context = getDebugInjectContext();
+
+      if (context === undefined || context.token === null) {
+        return;
+      }
+
+      if (typeof context.token !== 'object' && typeof context.token !== 'function') {
+        return;
+      }
```
It'd be helpful to have additional JSDoc for those exit conditions (to explain why it may happen).

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,121 @@
+
+      const {token, value, flags} = data as InjectedService;
+
+      let injectorKey: Injector|LView = context.injector;
```
Please add JSDoc here to explain why we have Injector or an LView.

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,121 @@
+      let injectorKey: Injector|LView = context.injector;
+
+      if (context.injector instanceof DebugNodeInjector) {
+        injectorKey = context.injector.lView!;
```
This would also benefit from  having a comment to explain why we use an LView as a key.

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,121 @@
+        injectorKey = context.injector.lView!;
+      }
+
+      if (!frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.has(injectorKey)) {
+        frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.set(
+            injectorKey, new WeakMap<Type<unknown>, InjectedService[]>());
+      }
+
+      if (!frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.get(injectorKey)!.has(
+              context.token)) {
+        frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.get(injectorKey)!.set(
+            context.token, []);
+      }
+
+      frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.get(injectorKey)!
+          .get(context.token)!.push({token, value, flags});
```
May be extract into a const for readability (const name is TBD, this is just to illustrate the idea)?
```suggestion
      const dependencyTracker = frameworkDIDebugData.injectorToInstantiatedTokenToDependencies;
      if (!dependencyTracker.has(injectorKey)) {
        // We see this key for the first time, create a WeakMap to collect deps.
        dependencyTracker.set(injectorKey, new WeakMap<Type<unknown>, InjectedService[]>());
      }

      if (!dependencyTracker.get(injectorKey)!.has(context.token)) {
        // This dependency is not yet in the WeakMap, add it.
        dependencyTracker.get(injectorKey)!.set(context.token, []);
      }

      // Register information about a token, its value and flags used for injection.
      dependencyTracker.get(injectorKey)!.get(context.token)!.push({token, value, flags});
```

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,121 @@
+          .get(context.token)!.push({token, value, flags});
+    }
+
+    if (type === InjectorProfilerEventType.Create) {
```
The next 2 `if`s would benefit from the same changes mentioned above.

AndrewKushnir commented on packages/core/src/render3/injector-profiler.ts:
```diff
@@ -0,0 +1,123 @@
+import type {SingleProvider} from '../di/provider_collection';
+import {Type} from '../interface/type';
+
+export const enum InjectorProfilerEventType {
```
I'd propose adding JSDocs for each symbol that we export from this file (in addition to JSDoc for each field, which is 👍), since they are key to understanding of this  subsystem.

AndrewKushnir commented on packages/core/src/render3/instructions/di.ts:
```diff
@@ -49,6 +51,16 @@ export function ɵɵdirectiveInject<T>(token: ProviderToken<T>, flags = InjectFl
   }
   const tNode = getCurrentTNode();
+
+  if (typeof ngDevMode === 'undefined' || ngDevMode) {
+    const value = getOrCreateInjectable<T>(tNode as any, lView, resolveForwardRef(token), flags);
+    injectorProfiler({
+      type: InjectorProfilerEventType.Inject,
+      data: {token: token as Type<unknown>, value: value as Type<unknown>, flags}
+    });
+    return value;
+  }
+
   return getOrCreateInjectable<T>(
       tNode as TDirectiveHostNode, lView, resolveForwardRef(token), flags);
```
Extract the code that gets the value outside of the `if` block:
```suggestion
  const value = getOrCreateInjectable<T>(tNode as any, lView, resolveForwardRef(token), flags);

  if (typeof ngDevMode === 'undefined' || ngDevMode) {
    injectorProfiler({
      type: InjectorProfilerEventType.Inject,
      data: {token: token as Type<unknown>, value: value as Type<unknown>, flags}
    });
    return value;
  }

  return value;
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -468,3 +478,167 @@ function assertDomElement(value: any) {
 }
+
+/**
+ *
```
Function description would be helpful here.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -468,3 +478,167 @@ function assertDomElement(value: any) {
+export function getDependenciesFromInstantiation(injector: Injector, token: Type<unknown>):
+    {instance: unknown; dependencies: InjectedService[]}|undefined {
+  const NOT_FOUND = {};
+  const instance = injector.get(token, NOT_FOUND, {self: true, optional: true});
```
Could you plz add a comment to explain why those flags are used?

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -468,3 +478,167 @@ function assertDomElement(value: any) {
+    return;
+  }
+
+  let injectorKey: Injector|LView = injector;
```
Having a quick comment to explain why we have an Injector or an LView here would be helpful.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -468,3 +478,167 @@ function assertDomElement(value: any) {
+  dependencies = dependencies.map(
+      dep => ({
+        ...dep,
+        providedIn: (typeof dep.value === 'object' ?
+                         frameworkDIDebugData.instanceToInjector.get(dep.value!) :
+                         frameworkDIDebugData.instanceToInjector.get(dep.token!)) ||
+            frameworkDIDebugData.instanceToInjector.get(dep.token!)
```
This is a bit unclear, could you please add a comment here too?

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -468,3 +478,167 @@ function assertDomElement(value: any) {
+  return {instance, dependencies};
+}
+
+export function getInjectorProviders(injector: Injector): ProviderRecord[] {
```
A couple comments:

- Could you please add a JSDoc here too?
- I think it'd be great to split the function into smaller fns and invoke them from here. That would also act as an implicit documentation of the logic (what happens when , etc).

AndrewKushnir replied:
I'll be happy to re-review this part once the function is split into smaller fns.

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -468,3 +478,167 @@ function assertDomElement(value: any) {
+  });
+}
+
+export function getInjectorParent(injector: Injector): Injector|null {
```
Same 2 comments: JSDoc + split this fn into smaller fns.

AndrewKushnir replied:
Btw, I think we can extract all those functions to a separate file like `injector_discovery_utils.ts`. Later on, we can also do some refactoring to create a new folder (like `packages/core/src/render3/util/discovery_utils/`) and put all relevant files there.

AndrewKushnir commented on packages/core/src/render3/di.ts:
```diff
@@ -700,6 +734,29 @@ function shouldSearchParent(flags: InjectFlags, isFirstHostTNode: boolean): bool
 }
 
+/**
+ * Debug wrapper around a NodeInjector.
```
Could you please add a bit more info to describe why we need it (which role does it play)?

AndrewKushnir reviewed (commented):
@AleksanderBodurri great work! 👍 

A couple high-level comments:

- it'd be helpful to add more comments into the code to cover function contracts as well as some tricky parts of the logic
- it'd be really awesome to minimize the footprint of the debug-related code in the main/production code, so it's easier to follow and see where dev mode starts and ends
- some functions would benefit from refactoring to split them into smaller fns

Thank you.

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,217 @@
+
+  setInjectorProfiler(({data, context, type}: InjectorProfilerEvent) => {
+    if (type === InjectorProfilerEventType.Inject) {
+      handleInjectEvent(context, data as InjectedService);
+    }
```
nit: we can do `return;`s here or `else if` here (no need to check other conditions once previous matches).

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,217 @@
+
+  const {token, value, flags} = data;
+
+  let diResolver: Injector|LView = context.injector;
```
Let's add a comment on why we may have both Injector and LView here.

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,217 @@
+    return;
+  }
+
+  // if token is a primitive type, ignore this event. We do this because we cannot keep track of
+  // non-primitive tokens in WeakMaps.
+  if (typeof context.token !== 'object' && typeof context.token !== 'function') {
+    return;
+  }
```
It looks like this condition is a bit out-of-sync with a similar one in another function above (no check for symbol).

AndrewKushnir replied:
You could also consider moving those checks to the function that invokes this one (it should be the same condition for all functions?).

AndrewKushnir commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,217 @@
+    return;
+  }
+
+  const environmentInjector: EnvironmentInjector|null =
+      context.injector.get(EnvironmentInjector, null, {optional: true});
+  if (environmentInjector === null) {
+    return;
+  }
```
This would also benefit from a quick comment in the code.

AndrewKushnir commented on packages/core/src/render3/injector_profiler.ts:
```diff
@@ -0,0 +1,202 @@
+   */
+  value: unknown;
+
+  /*
+    Flags that this service was injected with
```
```suggestion
  /**
   * Flags that this service was injected with
```

AndrewKushnir commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -0,0 +1,313 @@
+import {getParentInjectorIndex, getParentInjectorView, hasParentInjector} from './injector_utils';
+
+/**
+ *
```
I think we'd need some docs here to talk about the goal of this function.

AndrewKushnir commented on packages/core/src/render3/injector_profiler.ts:
```diff
@@ -0,0 +1,209 @@
+  injectorProfilerCallback = injectorProfiler;
+};
+
+const injectorProfiler = function(injectorEvent: InjectorProfilerEvent): void {
```
```suggestion
function injectorProfiler(injectorEvent: InjectorProfilerEvent): void {
```

AndrewKushnir replied:
This function (and other fns referenced in other parts of the code) would benefit from having JSDoc to describe what's the goal of the function (so that it's easy to access this info in IDE).

AndrewKushnir commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -0,0 +1,388 @@
+          continue;
+        }
+
+        const containerDef = getInjectorDef(container)!;
```
Could you plz add a quick comment to mention why we add a non-null assertion here (i.e. why injector def is always present at this location)?

AndrewKushnir commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -0,0 +1,388 @@
+
+        let isNextStepInPath = false;
+
+        // If the last container added to the path is in the imports of the current container
+        // unshift the current container into the path
```
The code in this function would also benefit from some extra comments, so that it's easier to restore the context later if needed.

AleksanderBodurri commented on packages/core/src/render3/framework-injector-profiler.ts:
```diff
@@ -0,0 +1,207 @@
+
+  // if token is a primitive type, ignore this event. We do this because we cannot keep track of
+  // non-primitive tokens in WeakMaps since they are not garbage collectable.
+  if (!isGarbageCollectable(context.token)) {
```
Fair enough. I mimicked the language used https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap for this function name but my feelings on the current name are not strongly held 🥁 

Looking at it closer, I think I prefer the way [ecma262](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#sec-canbeheldweakly) describes it: `canBeHeldWeakly`

JeanMeche commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -0,0 +1,503 @@
+        parentTNode as TElementNode | TContainerNode | TElementContainerNode, parentLView);
+  } else {
+    const chainedInjector = lView[INJECTOR] as ChainedInjector;
+    const injectorParent = (chainedInjector.injector as any).parent as Injector;
```
Can we do better than `any` here ?

AleksanderBodurri replied:
Would require an import of `OutletInjector` from `router` which we don't want. The larger issue here is that `core` should not have to concern itself with `router` concerns. It's unclear to me right now how we could work around this, but this logic will be refactored in the future with this in mind.

AndrewKushnir replied:
@AleksanderBodurri agree, we can refactor this later. Could you please add a TODO into the code for now, so that we can come back to it?

JeanMeche commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -0,0 +1,506 @@
+
+  let diResolver: LView;
+
+  if (injector instanceof DebugNodeInjector) {
+    diResolver = injector.lView;
+    return resolverToProviders.get(diResolver) ?? [];
+  }
+
+  diResolver = new DebugNodeInjector(injector).lView;
```
could be an if/else with a single return.

JeanMeche reviewed (commented):
Some minor nits from me 😊

AndrewKushnir commented on packages/core/src/render3/debug/injector_profiler.ts:
```diff
@@ -0,0 +1,224 @@
+}
+
+export function runInInjectorProfilerContext(
+    injector: Injector, token: Type<unknown>, callback: () => void): void {
+  const prevInjectContext = setInjectorProfilerContext({injector, token});
```
```suggestion
    injector: Injector, token: Type<unknown>, callback: () => void): void {
  !ngDevMode && throwError('Injector profiler should never be called in production mode');
  
  const prevInjectContext = setInjectorProfilerContext({injector, token});
```

AndrewKushnir commented on packages/core/src/render3/pipe.ts:
```diff
@@ -50,6 +52,11 @@ export function ɵɵpipe(index: number, pipeName: string): any {
 
   const pipeFactory = pipeDef.factory || (pipeDef.factory = getFactoryDef(pipeDef.type, true));
+
+  const prevInjectContext = setInjectorProfilerContext({
+    injector: createDebugNodeInjector(getCurrentTNode() as TTextNode, getLView()),
+    token: pipeDef.type
+  });
```
Missing `ngDevMode` wrapping (we should also need to make sure `setInjectorProfilerContext` and other functions used in the main code contain an `ngDevMode` check that prevents them from running in prod mode).

AleksanderBodurri committed:
test(core): unit tests for the injector profiler and injector debugging APIs

    Creates unit tests for the following APIs

    - setInjectorProfiler
    - getInjectorProviders
    - getInjectorResolutionPath
    - getDependenciesFromInjectable

    Modifies existing tests in

    - packages/examples/core/di/ts/injector_spec.ts
    - packages/core/test/render3/jit/declare_injectable_spec.ts
    - packages/core/test/render3/jit/declare_factory_spec.ts

    because they setup framework injector context manually.

    Exports setInjectorProfilerContext in packages/core/src/core_private_export.ts in order for use
    in the the modified tests above.

AndrewKushnir commented on packages/core/src/render3/pipe.ts:
```diff
@@ -50,6 +52,14 @@ export function ɵɵpipe(index: number, pipeName: string): any {
+  let previousInjectorProfilerContext: InjectorProfilerContext;
+  if (ngDevMode) {
+    previousInjectorProfilerContext = setInjectorProfilerContext({
+      injector: createDebugNodeInjector(getCurrentTNode() as TTextNode, getLView()),
```
Not sure that the `TTextNode` is the correct type here, should we just do `TNode` instead?

AleksanderBodurri replied:
I want to say in this code path, because this function is creating a pipe, the current TNode will always be of type `TTextNode` 🤔

If preferred I could just type this as the union type that `createDebugNodeInjector` takes in which is `TElementNode|TContainerNode|TElementContainerNode`

AndrewKushnir reviewed (approved):
@AleksanderBodurri LGTM (only 1 minor comment) 👍

AndrewKushnir reviewed (approved):
LGTM 👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:537930812:BASE:537930852:1685989143746:651ba609).

AndrewKushnir commented:
[TGP](https://test.corp.google.com/OCL:537930812:BASE:538040001:1686016916674:6ea37cc4).

AndrewKushnir commented:
[TGP #2](https://test.corp.google.com/OCL:537930812:BASE:547661590:1689214369542:ad7f5120).

AndrewKushnir commented:
FYI, the PR is **no longer blocked** on the internal Google tests (the necessary code is cleaned up).

alxhub commented on packages/core/src/render3/debug/framework_injector_profiler.ts:
```diff
@@ -0,0 +1,240 @@
+    context: InjectorProfilerContext, data: ProviderRecord): void {
+  const {resolverToProviders} = frameworkDIDebugData;
+
+  const diResolver = getDIResolver(context?.injector);
```
Same as above.

alxhub commented on packages/core/src/render3/debug/framework_injector_profiler.ts:
```diff
@@ -0,0 +1,240 @@
+  // In the NodeInjector case we do the same as above but first wrap the NodeInjector in a
+  // DebugNodeInjector
+  else if (injector instanceof NodeInjector) {
+    diResolver = new DebugNodeInjector(injector as NodeInjector).lView!;
```
I'm not super clear on the reasoning for creating a new `DebugNodeInjector` here when we could read `_lView` directly on the `NodeInjector`. Also, the cast shouldn't be necessary because of the `instanceof` check.

alxhub commented on packages/core/src/render3/debug/injector_profiler.ts:
```diff
@@ -0,0 +1,237 @@
+/**
+ * An object representing an event that is emitted through the injector profiler
+ */
+export interface InjectorProfilerEvent {
```
As mentioned above we should break this into specific interfaces and make `InjectorProfilerEvent` a type union of them, discriminated by the `type`.

alxhub commented on packages/core/src/render3/di.ts:
```diff
@@ -700,6 +734,38 @@ function shouldSearchParent(flags: InjectFlags, isFirstHostTNode: boolean): bool
+ * Having access to that data is useful because it helps identify injectors. Contrast NodeInjector
+ * to R3Injector which has a source string to help identify it.
+ */
+export class DebugNodeInjector implements Injector {
```
Do we really need this wrapper? Its various getters could just be implemented as utility functions that take a `NodeInjector`.

alxhub commented on packages/core/src/render3/di_setup.ts:
```diff
@@ -74,10 +75,21 @@ function resolveProvider(
+      const injector = createDebugNodeInjector(
+          tNode as TElementNode | TContainerNode | TElementContainerNode, lView);
+      runInInjectorProfilerContext(injector, token, () => {
+        providerFactory = providerToFactory(provider as SingleProvider);
```
Can you add a comment explaining what running `providerToFactory` in this context is actually doing? That operation doesn't perform any instantiation or interact with the DI profiler at all, as far as I can tell.

alxhub commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -0,0 +1,520 @@
+  const {resolverToTokenToDependencies} = getFrameworkDIDebugData();
+
+  let dependencies =
+      resolverToTokenToDependencies.get(diResolver)?.get?.(token as Type<unknown>) || [];
```
Minor nit: prefer `??` to `||` (yay modern JavaScript)

alxhub commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -0,0 +1,520 @@
+
+  dependencies = dependencies.map(dep => {
+    for (const injectorToCheck of getInjectorResolutionPath(injector)) {
+      const instance = injectorToCheck.get(
```
I think there might be a logic bug here.

It's entirely possible that `dep` was injected with the `SkipSelf` flag, meaning that even if the first injector in the resolution path had that token, it wouldn't have matched. But this search doesn't consider `dep.flags`, so it _would_ assume that `dep` came from the first injector.

We could take the flags into consideration here (including `Host`). Another approach would be to actually capture the injector from which a dependency came at the time that dependency was injected.

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:552143591:BASE:552143544:1690664286472:8efc6ab5).

AndrewKushnir commented:
FYI, [the Global Presubmit is "green"](https://fusion2.corp.google.com/presubmit/552143591/OCL:552143591:BASE:552153382:1690670364897:35cb6ac/targets).

AleksanderBodurri commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -353,6 +372,18 @@ export class R3Injector extends EnvironmentInjector {
+        // these are the only providers that do not go through the value hydration logic
+        // where this event would normally be emitted from.
+        if (isValueProvider(provider)) {
+          emitInstanceCreatedByInjectorEvent(provider.useValue);
```
Ahh good point. Lets address this and https://github.com/angular/angular/pull/48639#discussion_r1278097181 in a followup PR. I'll make a note of it so I don't forget
